### ISSUE TITLE
No schema available for azurerm_data_factory_integration_runtime_self_hosted

### ISSUE THREAD

**Issue by RichardFowles89** [terraform-providers/terraform-provider-azurerm#7581]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version
Terraform v  0.12.24
AzureRM v 2.16.0

### Affected Resource(s)

* azurerm_data_factory_integration_runtime_self_hosted

### Terraform Configuration Files

```hcl

data "azurerm_resource_group" "rg" {
  name = "my-rg"
}

data "azurerm_client_config" "client-config" {}

resource "azurerm_data_factory" "data-factory" {
  count               = var.enabled == true ? 1 : 0
  name                = "my-adf"
  location            = data.azurerm_resource_group.rg.location
  resource_group_name = data.azurerm_resource_group.rg.name

  identity {
    type = "SystemAssigned"
  }

  dynamic "vsts_configuration" {
    for_each = local.azure_devops_config
    iterator = azure_devops
    content {
      account_name    = azure_devops.value.account_name
      branch_name     = azure_devops.value.repo_branch_name
      project_name    = azure_devops.value.project_name
      repository_name = azure_devops.value.repo_name
      root_folder     = azure_devops.value.repo_root_folder
      tenant_id       = azure_devops.value.tenant_id
    }
  }

  tags = var.tags
}

resource "azurerm_data_factory_integration_runtime_managed" "managed_runtime" {
  count               = var.enabled == true && var.is_self_hosted == false ? 1 : 0
  name                = "my-managed-runtime"
  data_factory_name   = azurerm_data_factory.data-factory[0].name
  resource_group_name = data.azurerm_resource_group.rg.name
  location            = data.azurerm_resource_group.rg.location

  node_size = var.managed_runtime_node_size
}

resource "azurerm_data_factory_integration_runtime_self_hosted" "self_hosted_runtime" {
  count               = var.enabled == true && var.is_self_hosted == true ? 1 : 0
  name                = "my-self-hosted-runtime"
  resource_group_name = data.azurerm_resource_group.rg.name
  data_factory_name   = azurerm_data_factory.data-factory[0].name
}

data "azurerm_storage_account" "self_hosted_runtime_storage" {
  count               = var.enabled == true && var.is_self_hosted == true && var.use_existing_storage_account == true ? 1 : 0
  name                = "my-sa"
  resource_group_name = var.existing_storage_account_resource_group
}

resource "azurerm_storage_account" "self_hosted_runtime_storage" {
  count                    = var.enabled == true && var.is_self_hosted == true && var.use_existing_storage_account == false ? 1 : 0
  name                     = "my-sa"
  resource_group_name      = data.azurerm_resource_group.rg.name
  location                 = data.azurerm_resource_group.rg.location
  account_tier             = var.storage_account_tier
  account_replication_type = var.storage_account_replication_type

  tags = var.tags
}

resource "azurerm_storage_container" "self_hosted_runtime_container" {
  count                 = var.enabled == true && var.is_self_hosted == true ? 1 : 0
  name                  = "my-container"
  storage_account_name  = var.is_self_hosted == true ? data.azurerm_storage_account.self_hosted_runtime_storage[0].name : azurerm_storage_account.self_hosted_runtime_storage[0].name
  container_access_type = "private"
}

resource "azurerm_storage_blob" "self_hosted_runtime_installation_script" {
  count                  = var.enabled == true && var.is_self_hosted == true ? 1 : 0
  name                   = "my-script.ps1"
  storage_account_name   = var.is_self_hosted == true ? data.azurerm_storage_account.self_hosted_runtime_storage[0].name : azurerm_storage_account.self_hosted_runtime_storage[0].name
  storage_container_name = azurerm_storage_container.self_hosted_runtime_container[0].name
  type                   = "Block"
  source                 = "${path.module}/selfhosted/my-script.ps1"
}

resource "azurerm_virtual_machine_extension" "install_self_hosted_runtime" {
  count                = var.enabled == true && var.is_self_hosted == true ? 1 : 0
  name                 = "self-hosted-adf-integration-runtime-installation"
  virtual_machine_id   = var.self_hosted_vm_id
  publisher            = "Microsoft.Compute"
  type                 = "CustomScriptExtension"
  type_handler_version = "1.7"

  settings = <<SETTINGS
    {
        "fileUris": ["${azurerm_storage_blob.self_hosted_runtime_installation_script[0].url}"]
    }
SETTINGS
  protected_settings = <<PROTECTED_SETTINGS
    {
      "commandToExecute": "powershell -ExecutionPolicy Unrestricted -File my-script.ps1 '${azurerm_data_factory_integration_runtime_self_hosted.self_hosted_runtime[0].auth_key_1}'"
    }
  PROTECTED_SETTINGS
}

```

### Expected Behavior
terraform plan can be run against the code without an error. 

### Actual Behavior
After the initial deployment, all further attempts to run terraform plan produce the following error:

"Error: no schema available for module.datafactory.azurerm_data_factory_integration_runtime_self_hosted.self_hosted_runtime[0] while reading state; this is a bug in Terraform and should be reported"

### Steps to Reproduce

1. Deploy using terraform apply.
2. Then try to run terraform plan against the deployed environment. 

### Important Factoids
The code presented here is contained within a module. That is why the error refers to module.datafactory... etc

### References

* #0000

**Comment by tombuildsstuff** [comment]
hi @RichardFowles89 

Thanks for opening this issue.

The resource `azurerm_data_factory_integration_runtime_self_hosted` only recently became available in version 2.15 of the Azure Provider - where this is present in your state file this means that it was deployed via this version (or using a development version containing it).

The error message being returned from Terraform here says that the version of the provider being used doesn't support the `azurerm_data_factory_integration_runtime_self_hosted` resource, even though it's present in the state - which likely means the version of the Azure Provider being used is lower than when it was provisioned (presumably lower than version 2.15 of the Azure Provider).

To fix this you should be able to pin the version of the Azure Provider being used in the Provider block, for example:

```
provider "azurerm" {
  version = "= 2.17.0"

  features {}
}
```

and then re-run `terraform init` to ensure this version is used - which should solve this issue (and ensure that this version is used going forwards).

Since this should be fixed by updating the Terraform Configuration to include a constraint (and re-running `terraform init`) I'm going to close this issue for the moment - but if that doesn't work you should be able to get further assistance for this in [the Discuss forums](https://discuss.hashicorp.com).

Thanks!

**Comment by RichardFowles89** [comment]
Thank you @tombuildsstuff
